File: examples/data-science/003-hypothesis-testing/implementations/julia/hypothesis_testing.jl
#!/usr/bin/env julia

using Statistics
using Distributions
using Random

# Test result structure
struct TestResult
    statistic::Float64
    p_value::Float64
    df::Union{Float64, Nothing}
    reject_null::Bool
    alpha::Float64
end

# Normality test result
struct NormalityResult
    statistic::Float64
    p_value::Float64
    is_normal::Bool
end

# One-sample t-test
function one_sample_t_test(data::Vector{Float64}, μ₀::Float64; alpha::Float64=0.05)
    n = length(data)
    @assert n >= 2 "Sample size must be at least 2"
    
    x̄ = mean(data)
    s = std(data, corrected=true)
    se = s / sqrt(n)
    
    t_stat = (x̄ - μ₀) / se
    df = n - 1
    
    # Two-tailed p-value
    p_value = 2 * cdf(TDist(df), -abs(t_stat))
    
    TestResult(t_stat, p_value, df, p_value < alpha, alpha)
end

# Two-sample Welch's t-test (unequal variances)
function welch_t_test(sample1::Vector{Float64}, sample2::Vector{Float64}; alpha::Float64=0.05)
    n1, n2 = length(sample1), length(sample2)
    @assert n1 >= 2 && n2 >= 2 "Both samples must have at least 2 observations"
    
    x̄1, x̄2 = mean(sample1), mean(sample2)
    s1², s2² = var(sample1, corrected=true), var(sample2, corrected=true)
    
    se = sqrt(s1²/n1 + s2²/n2)
    t_stat = (x̄1 - x̄2) / se
    
    # Welch-Satterthwaite degrees of freedom
    df = (s1²/n1 + s2²/n2)^2 / ((s1²/n1)^2/(n1-1) + (s2²/n2)^2/(n2-1))
    
    # Two-tailed p-value
    p_value = 2 * cdf(TDist(df), -abs(t_stat))
    
    TestResult(t_stat, p_value, df, p_value < alpha, alpha)
end

# Paired t-test
function paired_t_test(sample1::Vector{Float64}, sample2::Vector{Float64}; alpha::Float64=0.05)
    n = length(sample1)
    @assert n == length(sample2) "Samples must have equal length for paired test"
    @assert n >= 2 "Sample size must be at least 2"
    
    differences = sample1 .- sample2
    d̄ = mean(differences)
    sd = std(differences, corrected=true)
    se = sd / sqrt(n)
    
    t_stat = d̄ / se
    df = n - 1
    
    # Two-tailed p-value
    p_value = 2 * cdf(TDist(df), -abs(t_stat))
    
    TestResult(t_stat, p_value, df, p_value < alpha, alpha)
end

# Mann-Whitney U test (Wilcoxon rank-sum test)
function mann_whitney_u_test(sample1::Vector{Float64}, sample2::Vector{Float64}; alpha::Float64=0.05)
    n1, n2 = length(sample1), length(sample2)
    @assert n1 >= 1 && n2 >= 1 "Both samples must have at least 1 observation"
    
    # Combine and rank
    combined = vcat(sample1, sample2)
    ranks = ordinalrank(combined)
    
    # Sum of ranks for each group
    R1 = sum(ranks[1:n1])
    R2 = sum(ranks[n1+1:end])
    
    # U statistics
    U1 = R1 - n1*(n1+1)/2
    U2 = R2 - n2*(n2+1)/2
    U = min(U1, U2)
    
    # Normal approximation for large samples
    μU = n1*n2/2
    σU = sqrt(n1*n2*(n1+n2+1)/12)
    
    # Continuity correction
    z = (U - μU + 0.5) / σU
    p_value = 2 * cdf(Normal(), -abs(z))
    
    TestResult(U, p_value, nothing, p_value < alpha, alpha)
end

# Wilcoxon signed-rank test
function wilcoxon_signed_rank_test(sample1::Vector{Float64}, sample2::Vector{Float64}; alpha::Float64=0.05)
    n = length(sample1)
    @assert n == length(sample2) "Samples must have equal length"
    @assert n >= 2 "Sample size must be at least 2"
    
    differences = sample1 .- sample2
    # Remove zeros
    differences = differences[differences .!= 0]
    n_eff = length(differences)
    
    if n_eff == 0
        return TestResult(0.0, 1.0, nothing, false, alpha)
    end
    
    # Rank absolute values
    abs_diffs = abs.(differences)
    ranks = ordinalrank(abs_diffs)
    
    # Sum of positive and negative ranks
    W_plus = sum(ranks[differences .> 0])
    W_minus = sum(ranks[differences .< 0])
    W = min(W_plus, W_minus)
    
    # Normal approximation
    μW = n_eff*(n_eff+1)/4
    σW = sqrt(n_eff*(n_eff+1)*(2*n_eff+1)/24)
    
    z = (W - μW + 0.5) / σW
    p_value = 2 * cdf(Normal(), -abs(z))
    
    TestResult(W, p_value, nothing, p_value < alpha, alpha)
end

# Chi-square test of independence
function chi_square_test(observed::Matrix{Float64}; alpha::Float64=0.05)
    rows, cols = size(observed)
    @assert rows >= 2 && cols >= 2 "Contingency table must be at least 2x2"
    
    total = sum(observed)
    row_sums = sum(observed, dims=2)
    col_sums = sum(observed, dims=1)
    
    # Expected frequencies
    expected = (row_sums * col_sums) / total
    
    # Chi-square statistic
    χ² = sum((observed .- expected).^2 ./ expected)
    
    # Degrees of freedom
    df = (rows - 1) * (cols - 1)
    
    # P-value
    p_value = 1 - cdf(Chisq(df), χ²)
    
    TestResult(χ², p_value, Float64(df), p_value < alpha, alpha)
end

# Shapiro-Wilk test for normality (simplified version)
function shapiro_wilk_test(data::Vector{Float64}; alpha::Float64=0.05)
    n = length(data)
    @assert n >= 3 "Sample size must be at least 3"
    
    # Sort data
    x = sort(data)
    x̄ = mean(x)
    
    # Calculate S²
    S² = sum((x .- x̄).^2)
    
    # Calculate W statistic (simplified)
    # For a full implementation, we'd need the Shapiro-Wilk coefficients
    # This is an approximation using normal probability plot correlation
    
    # Calculate expected normal order statistics
    m = n ÷ 2
    a = zeros(m)
    for i in 1:m
        p = (i - 0.375) / (n + 0.25)
        a[i] = quantile(Normal(), p)
    end
    
    # Calculate b
    b = 0.0
    for i in 1:m
        b += a[i] * (x[n+1-i] - x[i])
    end
    
    W = b^2 / S²
    
    # Approximate p-value (would need proper Shapiro-Wilk distribution)
    # Using a rough approximation based on sample size
    if n <= 50
        # Log-normal approximation for small samples
        μ = -1.2725 + 1.0521 * log(n)
        σ = 1.0308 - 0.2604 * log(n)
        z = (log(1 - W) - μ) / σ
        p_value = cdf(Normal(), z)
    else
        # Normal approximation for larger samples
        z = (W - 0.9) * sqrt(n) / 0.15
        p_value = cdf(Normal(), -abs(z))
    end
    
    NormalityResult(W, p_value, p_value > alpha)
end

# Anderson-Darling test for normality
function anderson_darling_test(data::Vector{Float64}; alpha::Float64=0.05)
    n = length(data)
    @assert n >= 2 "Sample size must be at least 2"
    
    # Standardize data
    x = sort(data)
    μ = mean(x)
    σ = std(x, corrected=true)
    z = (x .- μ) / σ
    
    # Calculate A² statistic
    S = 0.0
    for i in 1:n
        S += (2i - 1) * (log(cdf(Normal(), z[i])) + log(1 - cdf(Normal(), z[n+1-i])))
    end
    A² = -n - S/n
    
    # Adjusted statistic
    A²_star = A² * (1 + 0.75/n + 2.25/n^2)
    
    # Approximate p-value (simplified)
    if A²_star < 0.2
        p_value = 1 - exp(-13.436 + 101.14*A²_star - 223.73*A²_star^2)
    elseif A²_star < 0.34
        p_value = 1 - exp(-8.318 + 42.796*A²_star - 59.938*A²_star^2)
    elseif A²_star < 0.6
        p_value = exp(0.9177 - 4.279*A²_star - 1.38*A²_star^2)
    else
        p_value = exp(1.2937 - 5.709*A²_star + 0.0186*A²_star^2)
    end
    
    NormalityResult(A²_star, p_value, p_value > alpha)
end

# ANOVA (one-way)
function one_way_anova(groups::Vector{Vector{Float64}}; alpha::Float64=0.05)
    k = length(groups)
    @assert k >= 2 "Must have at least 2 groups"
    
    # Calculate group statistics
    n_groups = [length(g) for g in groups]
    n_total = sum(n_groups)
    
    group_means = [mean(g) for g in groups]
    grand_mean = sum(g -> sum(g), groups) / n_total
    
    # Sum of squares
    SS_between = sum(n_groups[i] * (group_means[i] - grand_mean)^2 for i in 1:k)
    SS_within = sum(sum((x - group_means[i])^2 for x in groups[i]) for i in 1:k)
    
    # Degrees of freedom
    df_between = k - 1
    df_within = n_total - k
    
    # Mean squares
    MS_between = SS_between / df_between
    MS_within = SS_within / df_within
    
    # F-statistic
    F_stat = MS_between / MS_within
    
    # P-value
    p_value = 1 - cdf(FDist(df_between, df_within), F_stat)
    
    TestResult(F_stat, p_value, Float64(df_between), p_value < alpha, alpha)
end

# Helper function for ranking
function ordinalrank(x::Vector{Float64})
    n = length(x)
    sorted_indices = sortperm(x)
    ranks = zeros(n)
    
    i = 1
    while i <= n
        j = i
        while j < n && x[sorted_indices[j+1]] == x[sorted_indices[i]]
            j += 1
        end
        
        # Average rank for ties
        avg_rank = (i + j) / 2
        for k in i:j
            ranks[sorted_indices[k]] = avg_rank
        end
        
        i = j + 1
    end
    
    return ranks
end

# Effect size calculations
function cohens_d(sample1::Vector{Float64}, sample2::Vector{Float64})
    n1, n2 = length(sample1), length(sample2)
    x̄1, x̄2 = mean(sample1), mean(sample2)
    s1, s2 = std(sample1, corrected=true), std(sample2, corrected=true)
    
    # Pooled standard deviation
    s_pooled = sqrt(((n1-1)*s1^2 + (n2-1)*s2^2) / (n1+n2-2))
    
    (x̄1 - x̄2) / s_pooled
end

# Main demonstration function
function main()
    Random.seed!(42)
    
    println("=== Hypothesis Testing in Julia ===\n")
    
    # Generate test data
    normal_data1 = randn(100) .* 2 .+ 5
    normal_data2 = randn(100) .* 2.5 .+ 5.5
    paired_data1 = randn(50) .* 1.5 .+ 10
    paired_data2 = paired_data1 .+ randn(50) .* 0.5 .+ 0.3
    
    # Parametric tests
    println("1. One-Sample T-Test")
    result = one_sample_t_test(normal_data1, 5.0)
    println("   t-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))")
    println("   Reject null: $(result.reject_null)\n")
    
    println("2. Welch's T-Test")
    result = welch_t_test(normal_data1, normal_data2)
    println("   t-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))")
    println("   df: $(round(result.df, digits=2))")
    println("   Cohen's d: $(round(cohens_d(normal_data1, normal_data2), digits=4))\n")
    
    println("3. Paired T-Test")
    result = paired_t_test(paired_data1, paired_data2)
    println("   t-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))\n")
    
    # Non-parametric tests
    println("4. Mann-Whitney U Test")
    result = mann_whitney_u_test(normal_data1, normal_data2)
    println("   U-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))\n")
    
    println("5. Wilcoxon Signed-Rank Test")
    result = wilcoxon_signed_rank_test(paired_data1, paired_data2)
    println("   W-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))\n")
    
    # Chi-square test
    println("6. Chi-Square Test")
    contingency = [10.0 15.0 8.0; 12.0 18.0 6.0; 8.0 10.0 14.0]
    result = chi_square_test(contingency)
    println("   χ²-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))")
    println("   df: $(Int(result.df))\n")
    
    # Normality tests
    println("7. Shapiro-Wilk Test")
    norm_result = shapiro_wilk_test(normal_data1)
    println("   W-statistic: $(round(norm_result.statistic, digits=4))")
    println("   p-value: $(round(norm_result.p_value, digits=4))")
    println("   Is normal: $(norm_result.is_normal)\n")
    
    println("8. Anderson-Darling Test")
    norm_result = anderson_darling_test(normal_data1)
    println("   A²-statistic: $(round(norm_result.statistic, digits=4))")
    println("   p-value: $(round(norm_result.p_value, digits=4))")
    println("   Is normal: $(norm_result.is_normal)\n")
    
    # ANOVA
    println("9. One-Way ANOVA")
    groups = [randn(30) .+ i for i in 1:3]
    result = one_way_anova(groups)
    println("   F-statistic: $(round(result.statistic, digits=4))")
    println("   p-value: $(round(result.p_value, digits=4))")
    println("   df: $(Int(result.df))\n")
    
    println("=== All hypothesis tests completed successfully ===")
end

# Run if executed directly
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: examples/data-science/003-hypothesis-testing/implementations/r/hypothesis_testing.R
#!/usr/bin/env Rscript

# Hypothesis Testing Implementation in R
# Sprint 25: Advanced Statistical Methods

set.seed(42)

# Test result structure
create_test_result <- function(statistic, p_value, df=NULL, reject_null, alpha=0.05) {
  list(
    statistic = statistic,
    p_value = p_value,
    df = df,
    reject_null = reject_null,
    alpha = alpha
  )
}

# Normality result structure
create_normality_result <- function(statistic, p_value, is_normal) {
  list(
    statistic = statistic,
    p_value = p_value,
    is_normal = is_normal
  )
}

# One-sample t-test
one_sample_t_test <- function(data, mu0, alpha=0.05) {
  stopifnot(length(data) >= 2)
  
  result <- t.test(data, mu = mu0, conf.level = 1 - alpha)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Two-sample Welch's t-test
welch_t_test <- function(sample1, sample2, alpha=0.05) {
  stopifnot(length(sample1) >= 2 && length(sample2) >= 2)
  
  result <- t.test(sample1, sample2, var.equal = FALSE, conf.level = 1 - alpha)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Paired t-test
paired_t_test <- function(sample1, sample2, alpha=0.05) {
  stopifnot(length(sample1) == length(sample2))
  stopifnot(length(sample1) >= 2)
  
  result <- t.test(sample1, sample2, paired = TRUE, conf.level = 1 - alpha)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Mann-Whitney U test (Wilcoxon rank-sum test)
mann_whitney_u_test <- function(sample1, sample2, alpha=0.05) {
  stopifnot(length(sample1) >= 1 && length(sample2) >= 1)
  
  result <- wilcox.test(sample1, sample2, exact = FALSE)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = NULL,
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Wilcoxon signed-rank test
wilcoxon_signed_rank_test <- function(sample1, sample2, alpha=0.05) {
  stopifnot(length(sample1) == length(sample2))
  stopifnot(length(sample1) >= 2)
  
  result <- wilcox.test(sample1, sample2, paired = TRUE, exact = FALSE)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = NULL,
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Chi-square test of independence
chi_square_test <- function(observed, alpha=0.05) {
  stopifnot(nrow(observed) >= 2 && ncol(observed) >= 2)
  
  result <- chisq.test(observed)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Shapiro-Wilk test for normality
shapiro_wilk_test <- function(data, alpha=0.05) {
  stopifnot(length(data) >= 3 && length(data) <= 5000)
  
  result <- shapiro.test(data)
  
  create_normality_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    is_normal = result$p.value > alpha
  )
}

# Anderson-Darling test (using nortest package if available)
anderson_darling_test <- function(data, alpha=0.05) {
  stopifnot(length(data) >= 2)
  
  # Check if nortest package is available
  if (requireNamespace("nortest", quietly = TRUE)) {
    result <- nortest::ad.test(data)
    statistic <- as.numeric(result$statistic)
    p_value <- result$p.value
  } else {
    # Simplified implementation without nortest
    n <- length(data)
    x <- sort(data)
    mu <- mean(x)
    sigma <- sd(x)
    z <- (x - mu) / sigma
    
    # Calculate A² statistic
    S <- 0
    for (i in 1:n) {
      S <- S + (2*i - 1) * (log(pnorm(z[i])) + log(1 - pnorm(z[n+1-i])))
    }
    A2 <- -n - S/n
    
    # Adjusted statistic
    A2_star <- A2 * (1 + 0.75/n + 2.25/n^2)
    
    # Approximate p-value
    if (A2_star < 0.2) {
      p_value <- 1 - exp(-13.436 + 101.14*A2_star - 223.73*A2_star^2)
    } else if (A2_star < 0.34) {
      p_value <- 1 - exp(-8.318 + 42.796*A2_star - 59.938*A2_star^2)
    } else if (A2_star < 0.6) {
      p_value <- exp(0.9177 - 4.279*A2_star - 1.38*A2_star^2)
    } else {
      p_value <- exp(1.2937 - 5.709*A2_star + 0.0186*A2_star^2)
    }
    
    statistic <- A2_star
  }
  
  create_normality_result(
    statistic = statistic,
    p_value = p_value,
    is_normal = p_value > alpha
  )
}

# Kolmogorov-Smirnov test
kolmogorov_smirnov_test <- function(data, alpha=0.05) {
  stopifnot(length(data) >= 2)
  
  result <- ks.test(data, "pnorm", mean = mean(data), sd = sd(data))
  
  create_normality_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    is_normal = result$p.value > alpha
  )
}

# Jarque-Bera test for normality
jarque_bera_test <- function(data, alpha=0.05) {
  n <- length(data)
  stopifnot(n >= 4)
  
  # Calculate moments
  mu <- mean(data)
  m2 <- mean((data - mu)^2)
  m3 <- mean((data - mu)^3)
  m4 <- mean((data - mu)^4)
  
  # Skewness and kurtosis
  skewness <- m3 / m2^(3/2)
  kurtosis <- m4 / m2^2 - 3
  
  # JB statistic
  JB <- n/6 * (skewness^2 + kurtosis^2/4)
  
  # P-value from chi-square distribution with 2 df
  p_value <- 1 - pchisq(JB, df = 2)
  
  create_normality_result(
    statistic = JB,
    p_value = p_value,
    is_normal = p_value > alpha
  )
}

# One-way ANOVA
one_way_anova <- function(groups, alpha=0.05) {
  stopifnot(length(groups) >= 2)
  
  # Combine groups into a single vector and create factor
  data_vec <- unlist(groups)
  group_factor <- factor(rep(1:length(groups), sapply(groups, length)))
  
  result <- aov(data_vec ~ group_factor)
  summary_result <- summary(result)[[1]]
  
  F_stat <- summary_result$`F value`[1]
  p_value <- summary_result$`Pr(>F)`[1]
  df_between <- summary_result$Df[1]
  
  create_test_result(
    statistic = F_stat,
    p_value = p_value,
    df = df_between,
    reject_null = p_value < alpha,
    alpha = alpha
  )
}

# Kruskal-Wallis test
kruskal_wallis_test <- function(groups, alpha=0.05) {
  stopifnot(length(groups) >= 2)
  
  # Combine groups into a single vector and create factor
  data_vec <- unlist(groups)
  group_factor <- factor(rep(1:length(groups), sapply(groups, length)))
  
  result <- kruskal.test(data_vec, group_factor)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Friedman test (for repeated measures)
friedman_test <- function(data_matrix, alpha=0.05) {
  stopifnot(nrow(data_matrix) >= 2 && ncol(data_matrix) >= 2)
  
  result <- friedman.test(data_matrix)
  
  create_test_result(
    statistic = as.numeric(result$statistic),
    p_value = result$p.value,
    df = as.numeric(result$parameter),
    reject_null = result$p.value < alpha,
    alpha = alpha
  )
}

# Effect size calculations
cohens_d <- function(sample1, sample2) {
  n1 <- length(sample1)
  n2 <- length(sample2)
  x1_bar <- mean(sample1)
  x2_bar <- mean(sample2)
  s1 <- sd(sample1)
  s2 <- sd(sample2)
  
  # Pooled standard deviation
  s_pooled <- sqrt(((n1-1)*s1^2 + (n2-1)*s2^2) / (n1+n2-2))
  
  (x1_bar - x2_bar) / s_pooled
}

# Eta squared for ANOVA
eta_squared <- function(groups) {
  data_vec <- unlist(groups)
  group_factor <- factor(rep(1:length(groups), sapply(groups, length)))
  
  result <- aov(data_vec ~ group_factor)
  summary_result <- summary(result)[[1]]
  
  SS_between <- summary_result$`Sum Sq`[1]
  SS_total <- sum(summary_result$`Sum Sq`)
  
  SS_between / SS_total
}

# Power analysis (simplified)
power_t_test <- function(n, effect_size, alpha=0.05, type="two.sample") {
  # Simplified power calculation for t-test
  if (type == "two.sample") {
    df <- 2*n - 2
    ncp <- effect_size * sqrt(n/2)
  } else {
    df <- n - 1
    ncp <- effect_size * sqrt(n)
  }
  
  critical_t <- qt(1 - alpha/2, df)
  power <- 1 - pt(critical_t, df, ncp) + pt(-critical_t, df, ncp)
  
  power
}

# Main demonstration function
main <- function() {
  cat("=== Hypothesis Testing in R ===\n\n")
  
  # Generate test data
  normal_data1 <- rnorm(100, mean = 5, sd = 2)
  normal_data2 <- rnorm(100, mean = 5.5, sd = 2.5)
  paired_data1 <- rnorm(50, mean = 10, sd = 1.5)
  paired_data2 <- paired_data1 + rnorm(50, mean = 0.3, sd = 0.5)
  
  # Parametric tests
  cat("1. One-Sample T-Test\n")
  result <- one_sample_t_test(normal_data1, 5.0)
  cat(sprintf("   t-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   Reject null: %s\n\n", result$reject_null))
  
  cat("2. Welch's T-Test\n")
  result <- welch_t_test(normal_data1, normal_data2)
  cat(sprintf("   t-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   df: %.2f\n", result$df))
  cat(sprintf("   Cohen's d: %.4f\n\n", cohens_d(normal_data1, normal_data2)))
  
  cat("3. Paired T-Test\n")
  result <- paired_t_test(paired_data1, paired_data2)
  cat(sprintf("   t-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n\n", result$p_value))
  
  # Non-parametric tests
  cat("4. Mann-Whitney U Test\n")
  result <- mann_whitney_u_test(normal_data1, normal_data2)
  cat(sprintf("   U-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n\n", result$p_value))
  
  cat("5. Wilcoxon Signed-Rank Test\n")
  result <- wilcoxon_signed_rank_test(paired_data1, paired_data2)
  cat(sprintf("   W-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n\n", result$p_value))
  
  # Chi-square test
  cat("6. Chi-Square Test\n")
  contingency <- matrix(c(10, 15, 8, 12, 18, 6, 8, 10, 14), nrow = 3, byrow = TRUE)
  result <- chi_square_test(contingency)
  cat(sprintf("   χ²-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   df: %d\n\n", as.integer(result$df)))
  
  # Normality tests
  cat("7. Shapiro-Wilk Test\n")
  norm_result <- shapiro_wilk_test(normal_data1)
  cat(sprintf("   W-statistic: %.4f\n", norm_result$statistic))
  cat(sprintf("   p-value: %.4f\n", norm_result$p_value))
  cat(sprintf("   Is normal: %s\n\n", norm_result$is_normal))
  
  cat("8. Anderson-Darling Test\n")
  norm_result <- anderson_darling_test(normal_data1)
  cat(sprintf("   A²-statistic: %.4f\n", norm_result$statistic))
  cat(sprintf("   p-value: %.4f\n", norm_result$p_value))
  cat(sprintf("   Is normal: %s\n\n", norm_result$is_normal))
  
  cat("9. Kolmogorov-Smirnov Test\n")
  norm_result <- kolmogorov_smirnov_test(normal_data1)
  cat(sprintf("   D-statistic: %.4f\n", norm_result$statistic))
  cat(sprintf("   p-value: %.4f\n", norm_result$p_value))
  cat(sprintf("   Is normal: %s\n\n", norm_result$is_normal))
  
  cat("10. Jarque-Bera Test\n")
  norm_result <- jarque_bera_test(normal_data1)
  cat(sprintf("   JB-statistic: %.4f\n", norm_result$statistic))
  cat(sprintf("   p-value: %.4f\n", norm_result$p_value))
  cat(sprintf("   Is normal: %s\n\n", norm_result$is_normal))
  
  # ANOVA
  cat("11. One-Way ANOVA\n")
  groups <- list(
    rnorm(30, mean = 1),
    rnorm(30, mean = 2),
    rnorm(30, mean = 3)
  )
  result <- one_way_anova(groups)
  cat(sprintf("   F-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   df: %d\n", as.integer(result$df)))
  cat(sprintf("   Eta squared: %.4f\n\n", eta_squared(groups)))
  
  cat("12. Kruskal-Wallis Test\n")
  result <- kruskal_wallis_test(groups)
  cat(sprintf("   H-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   df: %d\n\n", as.integer(result$df)))
  
  # Friedman test
  cat("13. Friedman Test\n")
  repeated_data <- matrix(rnorm(30*3), nrow = 30, ncol = 3)
  result <- friedman_test(repeated_data)
  cat(sprintf("   χ²-statistic: %.4f\n", result$statistic))
  cat(sprintf("   p-value: %.4f\n", result$p_value))
  cat(sprintf("   df: %d\n\n", as.integer(result$df)))
  
  # Power analysis
  cat("14. Power Analysis\n")
  power <- power_t_test(n = 50, effect_size = 0.5, alpha = 0.05, type = "two.sample")
  cat(sprintf("   Power for n=50, d=0.5: %.4f\n", power))
  power <- power_t_test(n = 100, effect_size = 0.5, alpha = 0.05, type = "two.sample")
  cat(sprintf("   Power for n=100, d=0.5: %.4f\n\n", power))
  
  cat("=== All hypothesis tests completed successfully ===\n")
}

# Run if executed directly
if (!interactive()) {
  main()
}